model/src/main/kotlin/utils/PurlUtils.kt

```diff
@@ -92,28 +93,44 @@ internal fun createPurl(
 ): String =
     buildString {
         append("pkg:")
-        append(type)
+        append(type.lowercase())
+        append('/')
 
         if (namespace.isNotEmpty()) {
+            append(namespace.trim('/').split('/').joinToString("/") { it.percentEncode() })
             append('/')
-            append(namespace.percentEncode())
+            append(name.trim('/').percentEncode())
+        } else {
+            append(name.percentEncode())
         }
 
-        append('/')
-        append(name.percentEncode())
+        if (version.isNotEmpty()) {
+            append('@')
 
-        append('@')
-        append(version.percentEncode())
+            // See https://github.com/package-url/purl-spec/blob/master/PURL-SPECIFICATION.rst#character-encoding which
+            // says "the '#', '?', '@' and ':' characters must NOT be encoded when used as separators".
+            val isChecksum = HashAlgorithm.VERIFIABLE.any { version.startsWith("${it.name.lowercase()}:") }
+            if (isChecksum) append(version) else append(version.percentEncode())
+        }
 
-        qualifiers.onEachIndexed { index, entry ->
+        qualifiers.filterValues { it.isNotEmpty() }.toSortedMap().onEachIndexed { index, entry ->
             if (index == 0) append("?") else append("&")
-            append(entry.key.percentEncode())
+
+            val key = entry.key.lowercase()
+            append(key)
+
             append("=")
-            append(entry.value.percentEncode())
+
+            // See https://github.com/package-url/purl-spec/blob/master/PURL-SPECIFICATION.rst#known-qualifiers-keyvalue-pairs.
+            if (key in KNOWN_KEYS) append(entry.value) else append(entry.value.percentEncode())
         }
```

Comment:
I've been trying to find the info in above linked docs without success. Can you give a hint 
how you arrived at KNOWN_KEYS being the set of qualifiers which need no encoding?

Reply:
I also struggled to find explicit docs about this, but without it the test suite fails.

Reply:
I'd prefer to stick to `KNOWN_QUALIFIER_KEYS`. It's shorter and does not hard-code any implications into its name.